plugins/node/opentelemetry-instrumentation-hapi/src/instrumentation.ts

```diff
@@ -337,15 +337,25 @@ export class HapiInstrumentation extends InstrumentationBase {
         const span = instrumentation.tracer.startSpan(metadata.name, {
           attributes: metadata.attributes,
         });
-        let res;
-        await api.context.with(
-          api.trace.setSpan(api.context.active(), span),
-          async () => {
-            res = await method(...params);
-          }
-        );
-        span.end();
-        return res;
+        try {
+          let res;
+          await api.context.with(
+            api.trace.setSpan(api.context.active(), span),
+            async () => {
+              res = await method(...params);
+            }
+          );
+          return res;
+        } catch (err) {
+          span.recordException(err);
+          span.setStatus({
+            code: api.SpanStatusCode.ERROR,
+            message: err.message,
+          });
+          throw err;
+        } finally {
+          span.end();
+        }
```

Comment:
`contex.with` passes return values through so the closure is not required here.
This or something similar should work:
```typescript
        try {
          return api.context.with(
            api.trace.setSpan(api.context.active(), span),
            () => method(...params)
          );
        } catch (err) {
          span.recordException(err);
          span.setStatus({
            code: api.SpanStatusCode.ERROR,
            message: err.message,
          });
          throw err;
        } finally {
          span.end();
        }
```
Additionally, `with` also takes arguments to be passed to the function, so it can be simplified even further:
```typescript
        try {
          return api.context.with(
            api.trace.setSpan(api.context.active(), span),
            method,
            ...params
          );
        } catch (err) {
          span.recordException(err);
          span.setStatus({
            code: api.SpanStatusCode.ERROR,
            message: err.message,
          });
          throw err;
        } finally {
          span.end();
        }
```
Test it out!
Docs: https://open-telemetry.github.io/opentelemetry-js/classes/_opentelemetry_context_async_hooks.asynchookscontextmanager.html#with

Reply:
Thanks! it looks much better now.